### PR TITLE
Add the Stokes Forces Implementation for the Surface Bands wave coupling method

### PR DESCRIPTION
**Pull request by alperaltuntas**
This PR adds the Stokes Forces implementation by @breichl and some additional refactoring and bug fixes. 

The summary of Stokes Forces changes (@breichl):

 - Introduction of the stokes forces based on the 3d Stokes drift profile. The 3d Stokes drift profile is obtained via the partitioned Stokes drift components exported from WW3. The added forces are the duS/dt term, the updated pressure gradient force, and the the vorticity modifications. For the duS/dt term, the increment tendency is computed whenever the Stokes drift is updated. This term is applied on the dynamics time step. The pressure gradient force is being calculated outside of the model's pressure solver, and  the Stokes-shear terms at cell centers and then integrating and taking gradients to estimate a pressure gradient tendency anomaly. This isn't the correct way of incorporating the pressure gradient force, so will be corrected soon. 
 
The summary of refactorings and bugfixes (@alperaltuntas) :

- Add the ability to import fields with ungridded dimensions in NUOPC cap.
- Add a more flexible and concise way of exporting partitioned stokes drift components from WW3. This is based on utilizing the NUOPC ungridded dims feature. The new way of exporting these fields is encapsulated with `if (cesm_coupled)` checks. However, when the anticipated unification of EMC and NCAR WW3 nuopc caps is accomplished, the old way of exporting the partitioned stokes drift components may be removed to get rid of duplicate code.
- Fixed the exact restart test failures by adding the 3d stokes drift field to the restart file. To do so,  minor refactoring is done in wave_interface module and calling driver modules.
- other bugfixes: remove unnecessary ampersands from variable declarations. enclose stokes vorticity diag computations with Stokes_VF conditional. Do not call get_Langmuir_Number if lamult is already provided via ww3. Fix openmp test by adding Stokes_VT to shr clause.

### CODE REVIEW

File: src/core/MOM_CoriolisAdv.F90 [hunk 1]
```diff
@@ -284,15 +293,41 @@ subroutine CorAdCalc(u, v, h, uh, vh, CAu, CAv, OBC, AD, G, GV, US, CS)
...
+        dudy(I,J) = ((u(I,j+1,k)-Waves%us_x(I,j+1,k))*G%dxCu(I,j+1) - &
+                     (u(I,j,k)-Waves%us_x(I,j,k))*G%dxCu(I,j))
+      enddo; enddo
+    else
+      do J=Jsq-1,Jeq+1 ; do I=Isq-1,Ieq+1
+        dvdx(I,J) = (v(i+1,J,k)*G%dyCv(i+1,J) - v(i,J,k)*G%dyCv(i,J))
+        dudy(I,J) = (u(I,j+1,k)*G%dxCu(I,j+1) - u(I,j,k)*G%dxCu(I,j))
+      enddo; enddo
+    endif
     do J=Jsq-1,Jeq+1 ; do I=Isq-1,Ieq+1
```
Comment (alperaltuntas) [inline src/core/MOM_CoriolisAdv.F90]: Just noting here that the following do loop (lines 331-334) needs to be removed, i think.

File: src/user/MOM_wave_interface.F90 [hunk 2]
```diff
@@ -1441,7 +1507,7 @@ subroutine CoriolisStokes(G, GV, dt, h, u, v, WAVES, US)
   integer :: i,j,k
 
   do k = 1, GV%ke
-    do j = G%jsc, G%jec
+do j = G%jsc, G%jec
```
Comment (breichl) [inline src/user/MOM_wave_interface.F90]: I think this was not intended.

File: src/user/MOM_wave_interface.F90 [hunk 3]
```diff
@@ -1537,6 +1746,51 @@ subroutine Waves_end(CS)
...
+  CS%Us_x(:,:,:) = 0.0
+  allocate(CS%Us_y(HI%isd:HI%Ied,HI%jsdB:HI%jedB,GV%ke))
+  CS%Us_y(:,:,:) = 0.0
+
+  call get_param(param_file,mdl,"WAVE_METHOD",wave_method_str, do_not_log=.true., default=NULL_STRING)
+
+  if (trim(wave_method_str)== trim(SURFBANDS_STRING)) then
+    vd(1) = var_desc("US_x", "m s-1", "3d zonal Stokes drift profile")
+    vd(2) = var_desc("US_y", "m s-1", "3d meridional Stokes drift profile")
+    call register_restart_field(CS%US_x(:,:,:), vd(1), .true., restart_CSp)
```
Comment (breichl) [inline src/user/MOM_wave_interface.F90]: The logical here has to be set to .false. to work in symmetric memory mode.  I do not understand why.

**Comment by alperaltuntas** [comment]
@gustavo-marques this is ready to be tested and merged.

**Comment by gustavo-marques** [comment]
@alperaltuntas, I have verified that this PR does not change answers for the MOM6-examples test suit w.r.t. the latest main (https://github.com/NCAR/MOM6/commit/9cb93044ef06a929171b5bd326a7c5c06aa72316). Were you able to run the CESM tests?

**Comment by alperaltuntas** [comment]
@gustavo-marques Yes, the aux_mom test suite passes.

**Pull request merged**